### PR TITLE
Allow linking empty .cmxa files on MSVC

### PR DESCRIPTION
**Pull request by dra27**
MSVC .lib format doesn't support having no `.obj` files in the library, so `ocamlopt -o foo.cmxa -a` generates `foo.cmxa` but not `foo.lib` (there's no error from the Microsoft Linker). The resulting `foo.cmxa` is un-linkable, since OCaml passes `foo.lib` on to the linker. This doesn't affect cc-based toolchains since `ar` rather more logically allows empty files!

This patch relaxes the requirement for `foo.lib` if `foo.cmxa` contains no units.

This is part of a fix for https://github.com/ocaml/stdlib-shims/issues/11

cc @db4

### CODE REVIEW

**Comment by dra27** [comment]
I intentionally haven't done anything to stop the linker being called when the .cmxa is created - at present, `lib` silently does nothing. If in future Microsoft were to make it an error (unlikely) then that'd need patching, but if the less unlikely change of supporting empty `.lib` files were added, then it would magically work (and the existing code would be there for compatibility with older releases)

File: asmcomp/asmlink.ml
Comment (xavierleroy) [inline asmcomp/asmlink.ml]: The house style is to have spaces around `::` and parentheses around tuples:
```
((info, file_name, crc) :: tolink, obj_name :: objfiles_rev)
```

File: asmcomp/asmlink.ml
Comment (xavierleroy) [inline asmcomp/asmlink.ml]: For symmetry I suggest
```
   let tolink = ...
   and objfiles_rev = ...
   in (tolink, objfiles_rev)
```
The two computations are independent, as far as I can see.

File: asmcomp/asmlink.ml
Comment (xavierleroy) [inline asmcomp/asmlink.ml]: It would make sense to just test `infos.lib_units = []`.  Even with non-MSVC toolchains, it makes sense to not pass an empty library to the linker, resulting in a shorter command line and less work for the linker.

As to testing that the file doesn't exist, this is to be commended as very prudent, but 1- it causes one extra system call per file to link, and 2- I can't imagine a scenario where object files would be added by hand to the library file that goes with an OCaml `.cmxa` file.

File: asmcomp/asmlink.ml
Comment (xavierleroy) [inline asmcomp/asmlink.ml:302]: I'm confused about the "rev" part, here and below.  This is a fold right, not a fold left, so it doesn't magically reverse lists.  E.g.
```
  List.fold_right (fun a l -> a :: l) l []
```
is the identity, not list reverse.

File: asmcomp/asmlink.ml
Comment (xavierleroy) [inline asmcomp/asmlink.ml]: For the reason mentioned above, I think the `List.rev` in the original code should be preserved.

**Review by xavierleroy — CHANGES_REQUESTED** [review]
There is something I don't understand with list reversals.  Plus some suggestions below.

File: asmcomp/asmlink.ml
Comment (dra27) [reply asmcomp/asmlink.ml]: I have no problem with doing that - the test I've put is just belt-and-braces to fix exactly the bug. The the `.a` file may have been subsequently altered (e.g. to add stub objects directly) - I've no knowledge of somewhere doing this, but do we definitely want to break some undefined behaviour while fixing this particular bug?

I did also consider whether it would be better to add a field to the .cmxa format which explicitly specifies the object to link - so on MSVC for an empty library, this field would explicitly be blank. In principle, .cmx format could do the same thing, although a `.o` and `.obj` file is always associated, even for an empty `.ml` file.

File: asmcomp/asmlink.ml
Comment (dra27) [reply asmcomp/asmlink.ml:302]: Oh man, that's embarrassing 🙁

**Comment by dra27** [comment]
Thanks, @xavierleroy! I'll push fixes later - what do you think of shifting the fix into .cmxa format?

File: asmcomp/asmlink.ml
Comment (gasche) [reply asmcomp/asmlink.ml]: In addition: I think this is one of the situations where it's nice to use `List.rev_map`.

Comment (dra27) [reply asmcomp/asmlink.ml]: Indeed - had I not been being an idiot about folds, I’d have switched it to that here!

**Comment by xavierleroy** [comment]
> what do you think of shifting the fix into .cmxa format?

I don't feel it would be a significant improvement.  But it would be a bigger change.  I like small changes :-)

**Comment by dra27** [comment]
Code corrected (I hope).

File: asmcomp/asmlink.ml
Comment (dra27) [reply asmcomp/asmlink.ml]: I (potentially temporarily) left this as-is. This change alters this PR from a bugfix to a potentially breaking change (albeit by using largely unspecified behaviour of the linker).

**Comment by damiendoligez** [comment]
Note that there is a similar problem on Macos (#6550, #1094) which was fixed in a completely different way. Is there a way to merge the two fixes?

**Comment by dra27** [comment]
No - the problem is not just that an empty `.lib` file isn't generated, it's also that the Microsoft Linker returns an error if you give it a `.lib` file containing no units (it's an error to try to remove the last one using the `LIB` tool as well)

File: asmcomp/asmlink.ml
Comment (Octachron) [reply asmcomp/asmlink.ml]: Am I reading right that the third part of the test is here to protect users that were creating a `.lib` file by hand  behind the back on the compiler on msvc? Do we really intend to support this use case?

Comment (dra27) [reply asmcomp/asmlink.ml]: @Octachron - it's more that I don't wish to intentionally break it in this PR. I believe there are some Mirage libraries which do this (although none which would hit this explicit case - i.e. an empty library with objects being added)

**Comment by xavierleroy** [comment]
> Note that there is a similar problem on Macos (#6550, #1094) which was fixed in a completely different way. Is there a way to merge the two fixes?

There is a way to simplify both fixes at the same time:
* At creation time, if the library has zero members, just create an empty `.a`/`.lib` file (`close_out (open_out f)`), don't call `ar`/`lib` at all.
* At link time, if the library has zero members, do not pass its `.a`/`.lib` file to the C linker.

That would be nicer than the current code + the current PR.

**Comment by dra27** [comment]
@xavierleroy - I'm puzzled, this PR does exactly your second point, with the only difference that the test is made specific to MSVC, doesn't it? (see https://github.com/ocaml/ocaml/pull/9011#discussion_r331781122 for why it's so specific). I don't at the moment see what I'm to change in _this_ PR.

For #1094, we'd just be simplifying the three calls in `Ccomp.macos_create_empty_archive`.

However, a null `foo.lib` is a _corrupt_ file - I'm not comfortable with the idea that we intentionally start generating files which the toolchain regards as corrupt, however weird the requirement of the toolchain may be. #1094 works around a bug where a toolchain doesn't generate a file it should be, which seems different.

Regardless, this PR should have at least one test, which I'm just doing...

**Comment by xavierleroy** [comment]
> - I'm puzzled, this PR does exactly your second point, with the only difference that the test is made specific to MSVC, doesn't it?

Yes.  But in order to simplify the macOS situation as well, linking of empty libraries should also be skipped on macOS.  At this point, let's just skip it on all platforms.

What I'm aiming at is a platform-independent solution to both the MSVC and the macOS issues.  Wouldn't this be nice?

**Comment by dra27** [comment]
So I can simply remove the `Config.ccomp_type = "msvc"` in the test in asmlink L207 above?

It turns out an empty .a is corrupt too:

```
dra27@summer:~$ touch foo.a ; ar t foo.a
ar: foo.a: File format not recognised
dra27@summer:~$ ar rc bar.a ; ar t bar.a
dra27@summer:~$ cat bar.a
!<arch>
```

**Comment by dra27** [comment]
OK, I think I've got hung on up on "simpler", but I think I follow where you're coming from. I certainly don't think we should be creating invalid files (so no zero-length files and no using other ways to create .lib files with no members).

MSVC forces us (and so build systems) to consider that `foo.cmxa` might not have `foo.a`/`foo.lib` so if we have MSVC not permitting empty archives and macOS unable to create them without trickery, then we could be entire platform independent - as you suggest - by not calling the archiver when there are no modules **including on Linux**.

**Comment by dra27** [comment]
I have verified that the test fails on trunk built with msvc64

**Comment by alainfrisch** [comment]
Not creating .a/.lib for empty libs seems natural enough, doesn't this complicate the life for build systems, who typically consider those .a/.lib files as dependencies for executables that link the libraries (and rightly so)?

**Comment by dra27** [comment]
@alainfrisch - it does complicate it, yes, but if that build system wishes to support MSVC properly then it *must* complicate it. That said, a simple build system need only worry about the .cmxa, since the only time a simple build system has to interact with the .a/.lib file is to install it.

**Comment by alainfrisch** [comment]
> it does complicate it, yes, but if that build system wishes to support MSVC properly then it must complicate it.

Well, another option would be to make sure the .a/.lib is not really empty, by adding a dummy (but valid) .o/.obj file to it.

> That said, a simple build system need only worry about the .cmxa, since the only time a simple build system has to interact with the .a/.lib file is to install it.

I don't understand.  If you define a rule to build an executable from a set of modules and libraries, the build system has to be aware of which files are used to build the resulting executable, which include the .a/.lib file for each library, no?

**Comment by dra27** [comment]
There's no need for a time-stamp based build system compiling an OCaml program ever to worry about `.a` and `.o` dependencies. Depending on `.cmx` and `.cmxa` is sufficient and ocamlopt hides all the details (cf. the venerable `OCamlMakefile` - executables depend on the .cmxa, it's only the install rules which pull in `.$(EXT_LIB)` files). Checksum based build systems are not simple :wink:

My strong opinions on this are that we should allow empty .cmxa files and we should not generate _invalid_ files in other formats (so no zero-byte .a or .lib files). I don't have a strong opinion as to which of the two possible ways to fix it should be done. The primary user of this feature is stdlib-shims, and Dune is already supports this (it was part of Dune 2.0). I lean more towards fixing a very small number of build systems once for something which should then never change vs a slightly dirtier hack (what you propose is basically what the macOS workaround does) which if it does affect a package will affect it in a weird way (conflict with the dummy name, or anything like that)

**Comment by xavierleroy** [comment]
The problem is not with the build system but with the "install" Makefile entries, `ocamlfind install`, and the like.  My suggestion to produce empty files with .a/.lib suffixes was so that these installers have something to copy.  Of course no linker will ever be given those fake .a/.lib files as argument.

**Comment by dra27** [comment]
Sure, but that's either one of two easy fixes in a `Makefile`, or an easy enough fix in `ocamlfind` for to understand that if `foo.a` can't be found and `foo.cmxa` is being installed then it's OK (and `ocamlfind` usually requires updating with every release of OCaml anyway). Is it just me who finds the idea of intentionally generating and installing a corrupt file distasteful? The Debian guys don't even like when we put valid extra bits in other people's file formats!

**Comment by alainfrisch** [comment]
> Checksum based build systems are not simple 😉

I was thinking about checksum based build systems indeed.  The build system might not know whether a library is empty or not (in particular is not under the control of the same build system), and anyway, it might be non-trivial logic to add in order to support this case.

**Comment by dra27** [comment]
Rebased onto latest trunk. I'm indescribably keen for the actual bugfix in this PR to be in 4.11 (that's the ability to run `echo 'print_endline "Hello, empty world!"' > bar.ml ; ocamlopt -a -o foo.cmxa ; ocamlopt -o bar foo.cmxa bar.ml` on msvc64 without error which is what the first commit fixes (and the second commit adds a test for).

The final commit goes further and alters the behaviour of `ocamlopt -a -o foo.cmxa` so that `foo.a` is not generated either. This latter part seems to have stalled the PR. Please can I either have approval to merge this PR as-is, or split it into two, with the first part being approved soon?

File: asmcomp/asmlink.ml
Comment (xavierleroy) [inline asmcomp/asmlink.ml]: Nitpick: the house style has one space on both sides of `::`.

**Review by xavierleroy — APPROVED** [review]
I'm pretty sure the files are now in the correct order, and I agree that not giving the linker archive files that were previously empty is a good thing to do.

I'm still a little worried about installation rules (in Makefiles or elsewhere) that systematically install a `.a`/`.obj` file for every library.  These will break, but perhaps the "empty library" case occurs rarely enough that we don't care.  I'm curious to know what Dune would do in this case, for example.

File: asmcomp/asmlink.ml
Comment (dra27) [reply asmcomp/asmlink.ml]: Fixed (and the one above it)

**Comment by dra27** [comment]
Dune's been fixed since 2.0.0 - see https://github.com/ocaml/dune/pull/2829

**Comment by dra27** [comment]
Thanks @xavierleroy, I'll merge and cherry-pick once CI replies

**Comment by dra27** [comment]
After a discussion with @Octachron, the plan is to merge this PR but only cherry-pick the first (bug-fix) commit to 4.11. The aforementioned Dune PR was written with the assumption that it was MSVC only, so it doesn't like Unix compilers not producing `.a` files. The fix for Dune is trivial (and I'm about to open the PR for it), but we think it's a little late in the release cycle for that change in 4.11.

So: 4.11.0 will be able to build stdlib-shims correctly on msvc and msvc64 and 4.12.0 gains the cleaned-up and simplified linker code.

**Pull request merged**
